src/realm/util/file.cpp

```diff
@@ -705,22 +705,27 @@ void File::prealloc(size_t size)
     // posix_fallocate() is not supported on MacOS or iOS
     fstore_t store = { F_ALLOCATEALL, F_PEOFPOSMODE, 0, static_cast<off_t>(new_size), 0 };
     int ret = fcntl(m_fd, F_PREALLOCATE, &store);
-    if (ret == -1) {
-        int err = errno;
-        std::string msg = get_errno_msg("fcntl() inside prealloc() failed: ", err);
-        throw OutOfDiskSpace(msg);
-    }
-
-    do {
-        ret = ftruncate(m_fd, new_size);
-    } while (ret == -1 && errno == EINTR);
-
-    if (ret != 0) {
-        int err = errno;
-        std::string msg = get_errno_msg("ftruncate() inside prealloc() failed: ", err);
-        throw OutOfDiskSpace(msg);
+    // On recent filesystems preallocation may fail.
+    if (ret != -1) {
+        // happy path! preallocation succesfull, so we'll use ftruncate to grow
+        // the file, since that is the fastest approach
+        do {
+            ret = ftruncate(m_fd, new_size);
+        } while (ret == -1 && errno == EINTR);
+
+        if (ret != 0) {
+            int err = errno;
+            std::string msg = get_errno_msg("ftruncate() inside prealloc() failed: ", err);
+            throw OutOfDiskSpace(msg);
+        }
+        return;
     }
-#elif REALM_ANDROID || defined(_WIN32)
+    // preallocation failed! fall through to slower, safe expansion of file
+    // by using writes. Preallocation failure is a "feature" in APFS, Apples
+    // new filesystem, as discussed here:
+    // https://www.mail-archive.com/filesystem-dev@lists.apple.com/msg00226.html
```

Comment:
Why fall through to this slow path rather than detecting that the file already had sufficient preallocated space (by using `fstat` as discussed in the post), and proceeding to call `ftruncate`?

Reply:
@bdash Ok, I see my misunderstanding here. I thought that the ftruncate() would "consume" the preallocation, so that after the call to ftruncate there'd be no preallocated memory left.

Reply:
Part of the problem is that we're preallocating too much space, due to a misunderstanding in how the API is used (see https://github.com/realm/realm-core/issues/3005#issuecomment-370083621). The other side of things is that the file system can, and does, preallocate more space than was requested, meaning free preallocated space will still remain after a `ftruncate`.